docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,449 @@
+.. _customizing-library-models-for-python:
+
+Customizing Library Models for Python
+=========================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+Python analysis can be customized by adding library models in data extension files.
+
+A data extension for Python is a YAML file of the form:
+
+.. code-block:: yaml 
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: <name of extensible predicate>
+      data:
+        - <tuple1>
+        - <tuple2>
+        - ...
+
+The CodeQL library for Python exposes the following extensible predicates:
+
+- **sourceModel**\(type, path, kind)
+- **sinkModel**\(type, path, kind)
+- **typeModel**\(type1, type2, path)
+- **summaryModel**\(type, path, input, output, kind)
+
+We'll explain how to use these using a few examples, and provide some reference material at the end of this article.
+
+Example: Taint sink in the 'fabric' package
+-------------------------------------------
+
+In this example, we'll show how to add the following argument, passed to **sudo** from the **fabric** package, as a command-line injection sink:
+
+.. code-block:: python
+
+  from fabric.api import sudo
+  sudo(cmd) # <-- add 'cmd' as a taint sink
+
+Note that this sink is already recognized by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sinkModel
+      data:
+        - ["fabric", "Member[api].Member[sudo].Argument[0]", "command-injection"]
+
+
+- Since we're adding a new sink, we add a tuple to the **sinkModel** extensible predicate.
+- The first column, **"fabric"**, identifies a set of values from which to begin the search for the sink.
+  The string **"fabric"** means we start at the places where the codebase imports the package **fabric**.
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+
+  - **Member[fabric]** selects accesses to the **fabric** package.
```

Comment:
```suggestion
  - **Member[api]** selects accesses to the **api** package.
```
This seems more in line with what the code is doing, but I'm also a bit confused, as I couldn't find any `api` package in the `fabric` repo.

Reply:
Indeed. It is mimicking this model of ours: https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/frameworks/Fabric.qll#L29-L48
but looking at the linked documentation, it seems to be called `operations` now.
We could
- update the example to use `operations` and stop claiming that our analysis already recognises this sink
- update the example and the existing modelling (and tests) to use `operations`
- ignore, possibly add "(outdated)" to the text

Reply:
I went for the second option...